apollo-federation/src/operation/selection_map.rs

```diff
@@ -0,0 +1,642 @@
+        }
+    }
+
+    /// Returns the number of selections in the map.
+    pub(crate) fn len(&self) -> usize {
+        self.selections.len()
+    }
+
+    /// Returns true if there are no selections in the map.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.selections.is_empty()
+    }
+
+    /// Returns the first selection in the map, or None if the map is empty.
+    pub(crate) fn first(&self) -> Option<&Selection> {
+        self.selections.first()
+    }
+
+    /// Computes the hash of a selection key; this borrows the whole map,
+    /// so mutating methods might want to use `hash_builder.hash_one` directly.
+    fn hash(&self, key: SelectionKey) -> u64 {
+        self.hash_builder.hash_one(key)
+    }
+
+    /// Returns true if the given key exists in the map.
+    pub(crate) fn contains_key(&self, key: SelectionKey) -> bool {
+        let hash = self.hash(key);
+        self.table
+            .find(hash, key_eq(&self.selections, key))
+            .is_some()
+    }
+
+    /// Returns true if the given key exists in the map.
+    pub(crate) fn get(&self, key: SelectionKey<'_>) -> Option<&Selection> {
+        let hash = self.hash(key);
+        let index = self.table.get(hash, key_eq(&self.selections, key))?;
+        Some(&self.selections[*index])
+    }
+
+    pub(crate) fn get_mut(&mut self, key: SelectionKey<'_>) -> Option<SelectionValue<'_>> {
+        let hash = self.hash(key);
+        let index = self.table.get(hash, key_eq(&self.selections, key))?;
+        Some(SelectionValue::new(&mut self.selections[*index]))
+    }
+
+    /// Insert a selection into the map.
+    fn raw_insert(&mut self, hash: u64, value: Selection) -> &mut Selection {
+        let index = self.selections.len();
+
+        // `insert` overwrites a selection without running `Drop`: because
+        // we only store an integer which is `Copy`, this works out fine
+        self.table.insert(hash, index, |existing| {
+            self.hash_builder.hash_one(self.selections[*existing].key())
+        });
+
+        self.selections.push(value);
+        &mut self.selections[index]
+    }
+
+    /// Resets and rebuilds the hash table.
+    ///
+    /// Preconditions:
+    /// - The table must have enough capacity for `self.selections.len()` elements.
+    fn rebuild_table_no_grow(&mut self) {
+        assert!(self.table.capacity() >= self.selections.len());
+        self.table.clear();
+        for (index, selection) in self.selections.iter().enumerate() {
+            let hash = self.hash(selection.key());
+            // SAFETY: Capacity is guaranteed by the assert at the top of the function
+            unsafe {
+                self.table.insert_no_grow(hash, index);
+            }
+        }
+    }
+
+    /// Decrements all the indices in the table starting at `pivot`.
+    fn decrement_table(&mut self, pivot: usize) {
+        // SAFETY: `self.table` must outlive `iter`
+        // None of the buckets can exit the function
+        unsafe {
```

Comment:
Since this is the first intro of `unsafe` into the crate, should we add the [`clippy::missing_safety_doc`](https://rust-lang.github.io/rust-clippy/master/index.html#/missing_safety_doc) safety lint to the crate-wide `warn` lints?

Reply:
Resolved as `hashbrown::HashTable`-based implementation does not require unsafe.